File: apps/api/src/middleware/tenant.ts
import { TENANT_HEADERS } from "@makyschool/shared/constants";
import { pool } from "../db/pool.js";
import type { NextFunction, Request, Response } from "express";
import { LRUCache } from "lru-cache";

export interface TenantRequest extends Request {
  schoolSlug?: string;
  schoolId?: string;
}

const schoolCache = new LRUCache<string, string>({
  max: 1000,
  ttl: 1000 * 60 * 5,
});

export function invalidateSchoolCache(slug: string) {
  schoolCache.delete(slug);
}

function isPublicRoute(path: string) {
  return (
    path.startsWith("/api/health") ||
    path.startsWith("/api/v1/health") ||
    path.startsWith("/api/auth") ||
    path.startsWith("/api/superadmin") ||
    path.startsWith("/api/webhooks")
  );
}

/** Resolves tenant from the school slug header and caches slug → id for 5 minutes. */
export function tenantMiddleware(
  req: TenantRequest,
  res: Response,
  next: NextFunction,
) {
  if (isPublicRoute(req.path)) {
    return next();
  }

  const schoolSlug = req.header(TENANT_HEADERS.SCHOOL_SLUG);

  if (!schoolSlug) {
    return res.status(400).json({
      error: "Missing tenant context",
      code: "TENANT_CONTEXT_REQUIRED",
    });
  }

  const cachedSchoolId = schoolCache.get(schoolSlug);
  if (cachedSchoolId) {
    req.schoolSlug = schoolSlug;
    req.schoolId = cachedSchoolId;
    return next();
  }

  pool
    .query<{ id: string }>("SELECT id FROM schools WHERE slug = $1 LIMIT 1", [
      schoolSlug,
    ])
    .then((result: { rows: Array<{ id: string }> }) => {
      const school = result.rows[0];

      if (!school) {
        res.status(404).json({ error: "School not found" });
        return;
      }

      schoolCache.set(schoolSlug, school.id);
      req.schoolSlug = schoolSlug;
      req.schoolId = school.id;
      next();
    })
    .catch(() => {
      res.status(500).json({ error: "Failed to resolve school tenant" });
    });
}

File: apps/api/app/middleware/tenant.py
from fastapi import Request, HTTPException, status
from typing import Optional
from app.db.pool import get_db_cursor

tenant_cache = {}

async def get_tenant_from_header(request: Request) -> Optional[dict]:
    school_slug = request.headers.get("x-school-slug")
    school_id = request.headers.get("x-school-id")
    
    if not school_slug:
        return None
    
    if school_slug in tenant_cache:
        return tenant_cache[school_slug]
    
    with get_db_cursor() as cursor:
        cursor.execute(
            "SELECT id, slug, name, status FROM schools WHERE slug = %s",
            (school_slug,)
        )
        school = cursor.fetchone()
        
        if school:
            tenant_cache[school_slug] = dict(school)
            return tenant_cache[school_slug]
    
    return None

def require_tenant(request: Request):
    school_slug = request.headers.get("x-school-slug")
    if not school_slug:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Missing tenant context"
        )
    return school_slug

def get_school_id(request: Request) -> int:
    school_id = request.headers.get("x-school-id")
    if not school_id:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Missing school ID in request"
        )
    try:
        return int(school_id)
    except ValueError:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid school ID"
        )
